include/execution.hpp

```diff
@@ -1692,6 +1676,111 @@ namespace std::execution {
       }
     } transfer {};
   } // namespace __transfer
+
+  inline namespace __on {
+    namespace __impl {
+      template <class R_>
+        struct __receiver_ref : receiver_adaptor<__receiver_ref<R_>> {
+          using R = __t<R_>;
+          R& __r_;
+          R&& base() && noexcept { return (R&&) __r_; }
+          const R& base() const & noexcept { return __r_; }
+        };
+
+      template <class Sch_, class S_>
+        struct __sender : sender_base {
+          using Sch = __t<Sch_>;
+          using S = __t<S_>;
+
+          Sch __sch_;
+          S __s_;
+
+          template <class R_>
+            struct __operation;
+
+          template <class R_>
+            struct __receiver : receiver_adaptor<__receiver<R_>> {
+              using R = __t<R_>;
+              __operation<R_>* __op_;
+              R&& base() && noexcept { return (R&&) __op_->__r_; }
+              const R& base() const & noexcept { return __op_->__r_; }
+
+              void set_value() && noexcept {
+                // cache this locally since *this is going bye-bye.
+                auto* op = __op_;
+                try {
+                  // This line will invalidate *this:
+                  start(op->__data_.template emplace<1>(__conv{
+                    [op]{
+                      return connect((S&&) op->__s_,
+                                     __receiver_ref<R_>{{}, op->__r_}); }
+                  }));
+                } catch(...) {
+                  set_error((R&&) op->__r_, current_exception());
+                }
+              }
+            };
+
+          template <class R_>
+            struct __operation {
+              using R = __t<R_>;
+
+              friend void tag_invoke(start_t, __operation& self) noexcept {
+                start(std::get<0>(self.__data_));
+              }
+
+              template <class S, class R>
+              __operation(Sch sch, S&& s, R&& r)
+                : __data_{in_place_index<0>, __conv{[&, this]{
+                    return connect(schedule(sch), __receiver<R_>{{}, this});
```

Comment:
This is not according to spec:
`Otherwise, constructs a sender s2. When s2 is connected with some receiver out_r, it results in an operation state op_state. When execution::start is called on op_state, it:`

Here, you are immediately  calling `execution::schedule` on the passed scheduler during construction of the operation state. However, it should only be called when `execution::start` is called, as that is the point where the receiver `r` is created.

Reply:
Yes, I think the spec is wrong. I'm planning to suggest this change. The spec is also wrong about the lifetime of op_state3.

Reply:
I am not sure whether it would be simpler to just store `sched` until we actually call `start` 

But yeah, both should work

Reply:
Another issue that comes from this is that you need the variant and the `__conv` dance.

If you store `sched` as a member and only call `execution::schedule` during `start` then you can immediately move the result from that operation to `execution::start` without having to store it.

Without the need to store two distinct types in the operation state you can simply go with a poor mans optional where you store the result from the `execution::connect` call

Reply:
Argh scratch that, I missed that we need to hold it..

Reply:
> Yes, I think the spec is wrong. I'm planning to suggest this change. The spec is also wrong about the lifetime of op_state3.

Are you talking about op_state2 which you construct here?  I would agree that it would be valid to drop it on the floor as you did once the `op_state3` comes along

Reply:
> Are you talking about `op_state2` which you construct here? I would agree that it would be valid to drop it on the floor as you did once the `op_state3` comes along

No, I mean `op_state3` which (somewhat confusingly) is constructed before `op_state2`. I should probably switch the names of the operation states and possibly restructure this section to avoid confusion.